src/hooks/useTodos.ts

```diff
@@ -0,0 +1,168 @@
+import {
+  useState,
+  useEffect,
+  Dispatch,
+  SetStateAction,
+  FormEvent,
+} from 'react';
+
+import {
+  USER_ID,
+  getTodos,
+  addTodo,
+  deleteTodo,
+  updateTodo,
+} from '../api/todos';
+
+import { useErrors } from './useErrors';
+
+import { Errors } from '../types/Errors';
+import { Todo } from '../types/Todo';
+
+type AddTodoAndSubmitFunction = (
+  e: FormEvent<HTMLFormElement>,
+  todo: Todo,
+  setTodo: Dispatch<SetStateAction<Todo>>,
+  tempTodo: Todo,
+) => void;
+
+export interface TodosContextValue {
+  todos: Todo[];
+  isLoading: boolean;
+  errors: Errors;
+  setTodos: Dispatch<SetStateAction<Todo[]>>;
+  setErrors: Dispatch<SetStateAction<Errors>>;
+  setIsLoading: Dispatch<SetStateAction<boolean>>;
+  loadingIds: number[];
+  triggerError: (errorKey: keyof Errors) => void;
+  addTodoAndSubmit: AddTodoAndSubmitFunction;
+  deletingTodo: (todoId: number) => void;
+  updatingTodo: (updatedTodo: Todo) => void;
+  toggleAllTodos: () => void;
+}
+
+export const useTodos = (): TodosContextValue => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingIds, setLoadingIds] = useState<number[]>([]);
+  const [errors, triggerError, setErrors] = useErrors({
+    titleError: false,
+    loadingError: false,
+    addTodoError: false,
+    updateTodoError: false,
+    deleteTodoError: false,
+  });
+
+  useEffect(() => {
+    const fetchTodos = async () => {
+      if (USER_ID) {
+        setIsLoading(true);
+        try {
+          const todosData = await getTodos();
+
+          setTodos(todosData);
+        } catch {
+          triggerError('loadingError');
+        } finally {
+          setIsLoading(false);
+        }
+      }
+    };
+
+    fetchTodos();
+  }, []);
+
+  const addTodoAndSubmit = async (
+    e: FormEvent<HTMLFormElement>,
+    todo: Todo,
+    setTodo: Dispatch<SetStateAction<Todo>>,
+    tempTodo: Todo,
+  ) => {
+    e.preventDefault();
+    if (!todo.title.trim()) {
+      triggerError('titleError');
+
+      return;
+    }
+
+    // const tempId = Date.now();
+    // const tempTodoItem = { ...todo, id: tempId };
+
+    // setTodos(prev => [...prev, tempTodoItem]);
```

Comment:
Remove commented. If you need that, leave a comment that you need it in the next iteration.